sw/device/tests/dif/dif_gpio_test.cc

```diff
@@ -0,0 +1,420 @@
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_irq_pin_read(&gpio_, 0, nullptr),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_irq_pin_read(nullptr, 0, nullptr),
+            kDifGpioResultInvalidArgument);
+
+  EXPECT_EQ(dif_gpio_irq_pin_clear(nullptr, 0), kDifGpioResultInvalidArgument);
+
+  EXPECT_EQ(dif_gpio_irq_masked_enable(nullptr, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+
+  EXPECT_EQ(dif_gpio_irq_masked_disable(nullptr, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+
+  EXPECT_EQ(dif_gpio_irq_trigger_masked_disable(nullptr, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+
+  EXPECT_EQ(dif_gpio_irq_trigger_masked_config(nullptr, 0xFFFFFFFF,
+                                               kDifGpioIrqEdgeRising),
+            kDifGpioResultInvalidArgument);
+}
+
+TEST_F(IrqTest, ForceSinglePin) {
+  for (uint32_t pin = 0; pin < 32; ++pin) {
+    EXPECT_WRITE32(GPIO_INTR_TEST_REG_OFFSET, {{pin, 1}});
+    EXPECT_EQ(dif_gpio_irq_pin_test(&gpio_, pin), kDifGpioResultOK);
+  }
+}
+
+TEST_F(IrqTest, ReadAllPins) {
+  uint32_t const val = 0xABABABAB;
+  EXPECT_READ32(GPIO_INTR_STATE_REG_OFFSET, val);
+  uint32_t irq_state = 0;
+  EXPECT_EQ(dif_gpio_irq_all_read(&gpio_, &irq_state), kDifGpioResultOK);
+  EXPECT_EQ(irq_state, val);
+}
+
+TEST_F(IrqTest, ReadSinglePin) {
+  for (uint32_t pin = 0; pin < 32; ++pin) {
+    for (const bool exp_irq_state : {true, false}) {
+      const uint32_t reg_val =
+          exp_irq_state ? AllZerosExcept(pin) : AllOnesExcept(pin);
+      EXPECT_READ32(GPIO_INTR_STATE_REG_OFFSET, reg_val);
+      bool irq_state = !exp_irq_state;
+      EXPECT_EQ(dif_gpio_irq_pin_read(&gpio_, pin, &irq_state),
+                kDifGpioResultOK);
+      EXPECT_EQ(irq_state, exp_irq_state);
+    }
+  }
+}
+
+TEST_F(IrqTest, ClearSinglePin) {
+  for (uint32_t pin = 0; pin < 32; ++pin) {
+    EXPECT_WRITE32(GPIO_INTR_STATE_REG_OFFSET, {{pin, 1}});
+    EXPECT_EQ(dif_gpio_irq_pin_clear(&gpio_, pin), kDifGpioResultOK);
+  }
+}
+
+TEST_F(IrqTest, MaskedEnable) {
+  uint32_t const val = 0xABABABAB;
+  EXPECT_READ32(GPIO_INTR_ENABLE_REG_OFFSET, 0x0);
+  EXPECT_WRITE32(GPIO_INTR_ENABLE_REG_OFFSET, val);
+
+  EXPECT_EQ(dif_gpio_irq_masked_enable(&gpio_, val), kDifGpioResultOK);
+}
+
+TEST_F(IrqTest, MaskedDisable) {
+  uint32_t const val = 0xABABABAB;
+  EXPECT_READ32(GPIO_INTR_ENABLE_REG_OFFSET, 0xFFFFFFFF);
+  EXPECT_WRITE32(GPIO_INTR_ENABLE_REG_OFFSET, ~val);
+
+  EXPECT_EQ(dif_gpio_irq_masked_disable(&gpio_, val), kDifGpioResultOK);
+}
+
+TEST_F(IrqTest, MaskedConfigTrigger) {
+  const uint32_t val = 0xABABABAB;
+
+  for (const dif_gpio_irq_t config :
+       {kDifGpioIrqEdgeRising, kDifGpioIrqEdgeFalling, kDifGpioIrqLevelLow,
+        kDifGpioIrqLevelHigh, kDifGpioIrqEdgeRisingFalling,
+        kDifGpioIrqEdgeRisingLevelLow, kDifGpioIrqEdgeFallingLevelHigh}) {
+    EXPECT_READ32(GPIO_INTR_CTRL_EN_RISING_REG_OFFSET, 0xFFFFFFFF);
```

Comment:
Can this block be split out in a separate `ExpectSomething` function? This pattern is repeated in several tests.

Reply:
I think that, if this is done, this test case should be unrolled from the loop, which would make it a bit easier to follow.

Reply:
Done.